silk/tools/otns_manager.py

```diff
@@ -0,0 +1,464 @@
+      if self.node is not None:
+        event = Event.status_event(message)
+        self.node.send_event(event.to_bytes())
+
+    get_extaddr_info_match = re.search(RegexType.GET_EXTADDR_RES.value, line)
+    if get_extaddr_info_match:
+      extaddr = get_extaddr_info_match.group(1)
+      if self.node is not None:
+        self.node.update_extaddr(int(extaddr, 16))
+
+  def subscribeHandle(self, sender, **kwargs):
+    """Handle messages from Publisher, a wpantund process.
+
+    Args:
+      sender (singal.Publisher): publisher of signal.
+      **kwargs (str): published signal.
+    """
+    line = kwargs["line"]
+    self.process_log_line(line)
+
+
+class OtnsManager(object):
+  """OTNS communication manager for a test case.
+
+  Attributes:
+    dispatcher_host (str): host address of OTNS dispatcher.
+    grpc_client (GRpcClient): OTNS gRPC client.
+    otns_nodes (List[OtnsNode]): OTNS nodes managed my this manager.
+    otns_node_map (Dict[ThreadDevBoard, OtnsNode]): map from device to
+      OTNS node.
+    otns_monitor_map (Dict[ThreadDevBoard, WpantundOtnsMonitor]): map from
+      device to OTNS monitor.
+    logger (Logger): logger for the manager class.
+  """
+
+  def __init__(self, dispatcher_host: str, logger: logging.Logger):
+    """Initialize an OTNS manager.
+
+    Args:
+      dispatcher_host (str): host address of OTNS dispatcher.
+      logger (logging.Logger): logger for the manager.
+    """
+    self.dispatcher_host = dispatcher_host
+    self.grpc_client = GRpcClient(
+        server_addr="{:s}:{:d}".format(dispatcher_host, GRPC_SERVER_PORT),
+        logger=logger.getChild("gRpcClient"))
+    self.otns_nodes = []
+    self.otns_node_map = {}
+    self.otns_monitor_map = {}
+
+    self.logger = logger
+
+  def add_node(self, node: ThreadDevBoard):
+    """Add a node to OTNS visualization.
+
+    Args:
+      node (ThreadDevBoard): node to add, with dev board properties.
+    """
+    if isinstance(node.device, HwModule):
+      try:
+        node_id = node.device.get_otns_vis_node_id()
+        vis_x, vis_y = node.device.get_otns_vis_location()
+        otns_node = OtnsNode(
+            node_id=node_id,
+            vis_x=vis_x,
+            vis_y=vis_y,
+            dispatcher_host=self.dispatcher_host,
+            dispatcher_port=DISPATCHER_PORT,
+            grpc_client=self.grpc_client,
+            logger=self.logger.getChild("otnsNode{:d}".format(node_id)))
+
+        node.otns_manager = self
+        self.otns_nodes.append(otns_node)
+        self.otns_node_map[node] = otns_node
+      except ValueError:
+        self.logger.error("Failed to get node OTNS properties.")
+    else:
+      self.logger.error("Trying to add a non HwModule node to manager.")
+
+  def remove_node(self, node: ThreadDevBoard):
+    """Remove a node from OTNS visualization.
+
+    Args:
+      node (ThreadDevBoard): node to remove, with dev board properties.
+    """
+    if node.otns_manager is self:
+      node.otns_manager = None
```

Comment:
should we just assert?
```suggestion
    assert(node.otns_manager is self)
    node.otns_manager = None
```

Reply:
Would it be possible that the node is already removed from the manager, but remove method is being called again? I think in that case the function should gracefully return instead of raising an exception.